AI guidelines (#176)
abtion/guidelines: technical_practices/agentic_programming.md

# PLAIbook — Abtion's AI-first work guidelines

This guide outlines good agentic development practices at Abtion.

Most of the information in here is merely recommendations. Devs free to deviate from these guidelines as they see fit. However! The output (commits, code, tests, etc.) should follow our best practices.

## Where we are, where we are headed

AI tools are improving rapidly, and so will this document.

> This document describes the now, and thus won't focus a lot on - for instance - autonomouse agents (We simply haven't established best practices yet)

Reviewer: ```suggestion
This document describes the now, and thus won't focus a lot on - for instance - autonomous agents (We simply haven't established best practices yet)
```

Reviewer: <img width="1254" height="1254" alt="Image" src="https://github.com/user-attachments/assets/0cc570d2-2eaa-402e-ac13-e7bd11e753fb" />

Author: Should I really remove this? - AI would never have come up with such a brilliant typo! 😄

### Where we are now

> - The developer is the architect (assisted by an agent with project context)

Reviewer: ```suggestion
- The developer is the architect with the vision, judgment, and responsibility. The agents execute the work with the relevant context.
```

- All work is primarily implemented locally by developers, mostly with agentic programming

There are experiments with:
- AI agents autonomously generating pull requests based on project tasks.

> - Standardizing task descriptions to make them more agent friendly (for both controlled and autonomous agents)

Reviewer: ```suggestion
- Standardizing many parts of the flow to make them more agent-friendly. Right now that includes task descriptions, autofixing agents, and gathering, summarizing, and processing large contexts from multiple sources, including meetings.
```

Reviewer: The reason that I add so many examples is that this is moving so rapidly that if we only keep that example, it will be outdated in a couple of months.

- Generating - the initial version of - minor projects based on designs from claude design.

### Where we - currently believe that we - are headed

- The developer is still the architect
- Initial boiler plate is generated from a design, based on a template
- Non-complex work is mostly done by autonomous AI agents (PRs reviewed/refactored if needed by human developers)
- Complex work is implemented locally by developers (still using agentic programming)

## Project setup checklist

Agents work best when the right context is available for them to load when needed.

Use the following checklist to set up context for your project:

> - [ ] `AGENTS.md` and symlink it from `CLAUDE.md`

Reviewer: ```suggestion
- [ ] `AGENTS.md` and symlink it from `CLAUDE.md`
- [ ] `CLAUDE.md` Should operate as an index file indicating where to find information in the whole repo, including documentation, skills, or any other source of context that is external to the repo.
```

Author: Not sure I agree on this, with this change it gets very fluffy whether or not AGENTS.md and CLAUDE.md are the same.

Reviewer: The symlink line stays, so they are the same.

The line is addressing what Claude.md should contain. It is the single most important context file, so defining what goes there is important.

- [ ] `docs/architecture.md`. Explains overall architecture of the project, reference in `AGENTS.md`
- [ ] `docs/decisions/*.md`. Information about important architectural decisions, referenced in `AGENTS.md`
- [ ] `.agents/skills/*`. Skills scoped to the project (for generally useful skills see: [how to create and share skills across projects](#how-to-create-and-share-skills-across-projects))
- [ ] `.agents/.mcp.json`. MCP server configuration, scoped to development of the specific project
- [ ] `.agents/skills/agent-setup/SKILL.md`. Sets up MCP servers and skills for the used agent (ask agent to `invoke ".agents/skills/agent-setup/SKILL.md"`)

Some of these files will be pre-populated by our templates, make sure to adjust them for the project.

> 

Reviewer: Maybe it's worth a note that we are experimenting on how to gather the best context for the project, and currently we are experimenting with Brewale and Palette as part of the process. 
The art of gathering the right context will take a big part and will keep evolving a lot during the rest of 2026.

Reviewer: ```suggestion
Gathering and sharing the right context is a big part of agentic programming, and this will keep evolving rapidly. We are experimenting with how to make useful tools, skills, and guidelines available for agents at the individual, project, and company level, while still leaving room for each project to define the business-specific context its agents need. At the time of writing, Brewale and Palette are part of that process.
```

Reviewer: I'm not really happy with the usage of the word "context" in my suggestion. 
It is a complicated one because I'm mixing up two things:
1. The context that is relevant in terms of the business
2. "How we operate" which is closer to guidelines or commands

Author: Is it intentional that this paragraph starts with "- ". It will render a bullet list with a single bullet.

I'm not sure that I think this paragraph belongs here as part of project setup. It seems more like a "where we are - where we are heading" kinda thing. Or otherwise this would have to be a lot more actionable.

Reviewer: Yeah, you are right. non-intentional. I was reading what was above.
Updated

Author: I applied this suggestion, but it didn't really fit in naturally in project checklist section.

Instead I moved it to the "Where we are now"-section and changed it a bit to make it better fit in:

> Another thing we are working on, which is not really an "experiment" but more an ongoing process is figuring out a good way to make useful tools, skills, and guidelines available for agents at the individual-, project-, and company level. We need to do this in a way that leaves room for each project to define the business-specific context its agents need. At the time of writing, Brewale and Palette are parts of that process.

> [!TIP]
> If a project is large or complex, split the contents of `docs/architecture.md` file into separate sections. Then reference each section in an overarching `docs/architecture.md`. This will make it easier for the agent to find the right information without scanning the entire repository.

### A note on monorepos

In monorepo projects, do the project setup checklist per-project, then add `AGENTS.md` (and `CLAUDE.md`) to the project root. It should explain the monorepo structure and list all the projects.

## Prerequisites for agentic programming

> - At least one coding agent installed, for instance:

Reviewer: ```suggestion
- At least one coding harness installed, for instance:
```

Author: My impression is that harness means everything around the model itself, including all the context and skills.

I did a quick roundtrip to see what each tool calls itself:

Claude: "Agentic coding tool".
Codex: "Coding agent"
Gemini: "AI agent"
OpenCode: "AI coding agent" (I love how it combines codex and gemini 😄)

Based on that I'd assume "coding agent" is more precise than "coding harness".

Reviewer: True. I did ask Claude before writing this. But you are right that they don't define themselves as harness.

## Working on multiple tasks in parallel

Since it takes time for agents to implement plans, it can feel inefficient to sit and wait for the agent to finish.

The obvious solution is to work on multiple tasks in parallel, and many tools are built around or directly support this ([Conductor](https://www.conductor.build/), [Cursor](https://cursor.com/), [herdr](https://herdr.dev/), [Zed](https://zed.dev/), [VS Code](https://code.visualstudio.com/))

It is recommended to have a separate [worktree](https://git-scm.com/docs/git-worktree) for each task to avoid conflicts. This is often something the tool will let you do out of the box.

It is our recommendation to be thoughtful about working on things in parallel. Based on an internal survey (June 2026), while devs tend to answer that parallel work makes them more productive, the conclusion is not very clear. Also no devs found it ideal to work on more than at most 3 tasks at a time.

> [!TIP]

> > Try if you can spent the "idle" time on thinking about or researching the task at hand (or working on something that supports the task).

Reviewer 2: ```suggestion
> Try if you can spend the "idle" time on thinking about or researching the task at hand (or working on something that supports the task).
```

## Guidelines for AI-generated PRs

On some projects we are experimenting with autonomous agents, letting them create and submit PRs on their own.

> While this might save us some time, it also introduces security / economic concerns. Agents can easily get stuck in loops, and they are vulnerable to prompt injections.

Reviewer: The phrasing seems opinionated.

```suggestion
As we experiment with new ways of working, we have to be aware of the security and economic concerns: agents can become stuck in loops, interpret instructions too literally, and are vulnerable to prompt injection.
```

Author: Yeah, I was also not entirely happy with this sentence. The suggestion I however feel is becoming too wordy.

Could it instead be:
> To avoid security and economic concerns, such as: agents getting stuck in loops, instructions interpretted too literally, and prompt injections, it is a requirement that autonomous agents:
  [the list below]

Reviewer: Sounds good!

Though spinning up many agents might seem like an obvious productivity boost, be thoughtful about working on things in parallel. Based on an internal survey (June 2026), while devs tend believe that parallel work make them more productive, the conclusion is not clear. Also all devs found it ideal to work on 1-3 tasks at a time.

> [!TIP]
> Instead of grabbing another task while the agent churns, try reflecting on the task at hand. There might be other ways to advance that task, avoiding excessive context switching.

## Guidelines for AI-generated PRs

On some projects we are experimenting with autonomous agents, letting them create and submit PRs on their own.

To avoid security and economic concerns, such as: agents getting stuck in loops, instructions interpretted too literally, and prompt injections, it is a requirement that autonomous agents:

- Do not have access to confidential keys and or tokens
- Are not able to commit directly to the main branch
- Can only submit draft PRs
- Have a hard TTL (e.g. they are killed if they take too long)

## Prompt engineering basics and Abtion-specific patterns

> Is this still needed?

Reviewer: Nah. Let's remove it for now. It makes the document look incomplete.
